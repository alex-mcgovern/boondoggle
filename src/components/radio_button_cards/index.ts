export { RadioButtonCards } from "./radio_button_cards.comp";
export type {
  RadioButtonCardsProps,
  RadioButtonCardShape,
} from "./radio_button_cards.comp";
