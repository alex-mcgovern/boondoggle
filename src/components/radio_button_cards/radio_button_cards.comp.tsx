import clsx from "clsx";
import { forwardRef, useMemo } from "react";

import { getTheme } from "../../styles/color_palette.css";
import { Box } from "../box";
import { InputDescription } from "../input_description";
import { InputErrorMessage } from "../input_error_message";
import { Label } from "../label";
import {
  getGroupLabelStyles,
  radioButtonInputStyles,
  radioButtonLabelStyles,
} from "./radio_button_cards.styles.css";

import type { SprinklesArgs } from "../../styles/utils/get_sprinkles.css";
import type { LabelledElementCustomisation } from "../../types";
import type { HTMLProps, ReactNode, Ref } from "react";

/** ----------------------------------------------------------------------------- */

type RadioButtonInputProps = Omit<
  HTMLProps<HTMLInputElement>,
  "className" | "id" | "name" | "required" | "type" | "value"
>;

/** ----------------------------------------------------------------------------- */

export type RadioButtonCardShape = {
  adornment: ReactNode;
  body?: string;
  checked?: boolean;
  tag?: string;
  title: string;
  value: string;
};

/** ----------------------------------------------------------------------------- */

type RadioButtonCardProps = RadioButtonCardShape & {
  name: string;
  onChange?: (value: string) => void;
  radioButtonInputProps?: RadioButtonInputProps;
  required?: boolean;
};

export function RadioButtonCard({
  adornment,
  body,
  checked,
  name,
  onChange,
  radioButtonInputProps,
  required,
  title,
  value,
}: RadioButtonCardProps) {
  return (
    <Box
      as="label"
      className={radioButtonLabelStyles}
      htmlFor={`${name}-${value}`}
      position="relative"
    >
      <input
        className={radioButtonInputStyles}
        defaultChecked={checked}
        id={`${name}-${value}`}
        name={name}
        onChange={() => {
          onChange?.(value);
        }}
        required={required}
        type="radio"
        value={value}
        {...radioButtonInputProps}
      />
      <Box
        alignItems="center"
        display="flex"
        flexDirection="column"
        justifyContent="center"
        textAlign="center"
      >
        {adornment && <Box marginBottom="spacing_2">{adornment}</Box>}
        <Box alignItems="center" display="flex" gap="spacing_1">
          <Box color="text_high_contrast">{title}</Box>
        </Box>
        <Box color="text_low_contrast" fontStyle="body_xs">
          {body}
        </Box>
      </Box>
    </Box>
  );
}

/** ----------------------------------------------------------------------------- */

export type RadioButtonCardsProps = SprinklesArgs &
  LabelledElementCustomisation & {
    defaultValue?: string | number;
    description?: string;
    errorMessage?: string;
    id: string;
    invalid?: boolean;
    isLabelVisible?: boolean;
    items: Array<RadioButtonCardShape>;
    label: string;
    labelTooltip?: string;
    name: string;
    onChange?: (value: string) => void;
    radioButtonInputProps?: RadioButtonInputProps;
    required?: boolean;
    value?: string;
  };

export const RadioButtonCards = forwardRef(
  (
    {
      defaultValue,
      description,
      errorMessage,
      id,
      invalid,
      isLabelVisible,
      items,
      label,
      labelProps,
      labelTooltip,
      name,
      onChange,
      radioButtonInputProps,
      required,
      value,
      wrapperProps,
      ...rest
    }: RadioButtonCardsProps,
    ref: Ref<HTMLDivElement>
  ) => {
    const controlledItems: Array<RadioButtonCardShape> = useMemo(() => {
      if (!Array.isArray(items) || items.length < 1) {
        return [];
      }
      return items.map((item) => {
        return {
          ...item,
          checked: defaultValue === item.value || value === item.value,
        };
      });
    }, [defaultValue, items, value]);

    /** --------------------------------------------- */

    if (!Array.isArray(controlledItems) || controlledItems.length < 1) {
      return null;
    }

    /** --------------------------------------------- */

    return (
      <Box
        className={clsx({ [getTheme({ colorOverlay: "red" })]: invalid })}
        {...wrapperProps}
        ref={ref}
      >
        {label && id && (
          <Label
            className={getGroupLabelStyles({ isLabelVisible })}
            htmlFor={id}
            id={`label-${id}`}
            label={label}
            labelTooltip={labelTooltip}
            {...labelProps}
          />
        )}
        <Box
          aria-labelledby={`label-${id}`}
          as="fieldset"
          display="flex"
          flexWrap="wrap"
          gap="spacing_1"
          id={id}
          name={name}
          {...rest}
        >
          {controlledItems.map((item) => {
            return (
              <RadioButtonCard
                adornment={item.adornment}
                body={item.body}
                checked={item.checked}
                key={item.title}
                name={name}
                onChange={onChange}
                radioButtonInputProps={radioButtonInputProps}
                required={required}
                title={item.title}
                value={item.value}
              />
            );
          })}
        </Box>
        {invalid && errorMessage && (
          <InputErrorMessage message={errorMessage} />
        )}
        {description && !invalid && (
          <InputDescription description={description} />
        )}
      </Box>
    );
  }
);

/** ----------------------------------------------------------------------------- */
