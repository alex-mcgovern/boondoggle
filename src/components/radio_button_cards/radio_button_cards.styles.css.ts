import { style } from "@vanilla-extract/css";
import { recipe } from "@vanilla-extract/recipes";

import { a11yFocus, a11yFocusStyleRule } from "../../styles/common/a11y.css";
import { vars } from "../../styles/theme.css";
import { createAccessibleTransition } from "../../styles/utils/create_accessible_transition";
import { getSprinkles } from "../../styles/utils/get_sprinkles.css";

/** ----------------------------------------------------------------------------- */

export const radioButtonListItemStyles = style([
  getSprinkles({ marginBottom: "spacing_1" }),
  {
    listStyle: "none",
  },
]);

/** ----------------------------------------------------------------------------- */

export const getGroupLabelStyles = recipe({
  base: [],
  variants: {
    isLabelVisible: {
      false: {
        display: "none",
      },
      true: {},
    },
  },
});

/** ----------------------------------------------------------------------------- */

export const radioButtonInputStyles = style({
  position: "absolute",
  right: vars.spacing.spacing_2,
  selectors: {
    [`&:focus, &:focus-visible`]: {
      outline: "none",
    },
  },
  top: vars.spacing.spacing_2,
});

/** ----------------------------------------------------------------------------- */

export const radioButtonLabelStyles = style([
  a11yFocus,
  getSprinkles({
    alignItems: "center",
    aspectRatio: "square",
    border: "border_default",
    borderRadius: "sm",
    display: "flex",
    gap: "spacing_2",
    justifyContent: "center",
    padding: "spacing_2",
    width: "100%",
  }),
  {
    ...createAccessibleTransition({
      transition: `border-color ${vars.transitionDuration.short} ease, background ${vars.transitionDuration.short} ease`,
    }),
    accentColor: vars.color.button_default,
    maxWidth: "12rem",
    selectors: {
      "&:has(input:checked)": {
        background: vars.color.tint_default,
        border: `1px solid ${vars.color.border_active}`,
      },
      "&:has(input:focus)": {
        ...a11yFocusStyleRule,
        border: `1px solid ${vars.color.button_default}`,
      },
      "&:hover": {
        background: vars.color.tint_default,
        border: `1px solid ${vars.color.border_active}`,
        cursor: "pointer",
      },
    },
  },
]);

/** ----------------------------------------------------------------------------- */

export const radioButtonStyles = style({
  pointerEvents: "none",
});
