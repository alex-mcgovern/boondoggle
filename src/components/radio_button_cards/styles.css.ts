import { style } from "@vanilla-extract/css";
import { recipe } from "@vanilla-extract/recipes";

import { a11yFocus, a11yFocusStyleRule } from "../../styles/common/a11y.css";
import { vars } from "../../styles/theme.css";
import { getSprinkles } from "../../styles/utils/get_sprinkles.css";

/** ----------------------------------------------------------------------------- */

export const radioButtonListItemStyles = style([
  getSprinkles({ marginBottom: "spacing_2" }),
  {
    listStyle: "none",
  },
]);

/** ----------------------------------------------------------------------------- */

export const radioButtonInputStyles = style({
  selectors: {
    [`&:focus, &:focus-visible`]: {
      outline: "none",
    },
  },
});

/** ----------------------------------------------------------------------------- */

export const getRadioButtonCardStyles = recipe({
  base: [
    a11yFocus,
    getSprinkles({
      alignItems: "center",
      borderRadius: "sm",
      display: "flex",
      gap: "spacing_2",
      padding: "spacing_2",
      width: "100%",
    }),
    {
      selectors: {
        "&:has(input:checked)": {
          background: vars.color.tint_default,
        },
        "&:has(input:focus)": {
          ...a11yFocusStyleRule,
        },
        "&:hover": {
          background: vars.color.tint_default,
          cursor: "pointer",
        },
      },
    },
  ],
});

/** ----------------------------------------------------------------------------- */

export const radioButtonStyles = style({
  pointerEvents: "none",
});
