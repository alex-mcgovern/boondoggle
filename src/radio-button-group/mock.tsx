import type * as React from "react";

import type { RadioButton } from "../radio-button";

import { LOREM } from "../../mocks/LOREM.mock";

export const RADIO_BUTTONS_MOCK: Array<
	React.ComponentProps<typeof RadioButton>
> = [
	{
		description: LOREM.text_xxs,
		name: "radio_item_1",
		title: "Radio item 1",
		value: "radio_item_1",
	},
	{
		description: LOREM.text_xxs,
		name: "radio_item_2",
		title: "Radio item 2",
		value: "radio_item_2",
	},
	{
		description: LOREM.text_xxs,
		name: "radio_item_3",
		title: "Radio item 3",
		value: "radio_item_3",
	},
];
