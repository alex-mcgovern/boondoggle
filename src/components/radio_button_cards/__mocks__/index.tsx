import {
  faArrowRightArrowLeft,
  faBank,
  faShapes,
} from "@fortawesome/pro-light-svg-icons";

import { LOREM } from "../../../../mocks/LOREM.mock";
import { Icon } from "../../icon";

export const RADIO_BUTTON_CARDS_MOCK = [
  {
    adornment: <Icon icon={faBank} size="xl" />,
    body: LOREM.text_md,
    title: "Item 1",
    value: "item_1",
  },
  {
    adornment: <Icon icon={faArrowRightArrowLeft} size="xl" />,
    body: LOREM.text_md,
    title: "Item 2",
    value: "item_2",
  },
  {
    adornment: <Icon icon={faShapes} size="xl" />,
    body: LOREM.text_md,
    title: "Item 3",
    value: "item_3",
  },
];
