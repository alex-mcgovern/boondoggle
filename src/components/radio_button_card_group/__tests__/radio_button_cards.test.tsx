/**
 * @jest-environment jsdom
 */
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { RadioButtonCardGroup } from "..";
import { RADIO_BUTTON_CARDS_MOCK } from "../__mocks__/radio_button_cards.mock";

import type { RadioButtonCardGroupProps } from "..";

const onChangeMock = jest.fn();

const PROPS: RadioButtonCardGroupProps = {
    id: "test",
    items: RADIO_BUTTON_CARDS_MOCK,
    label: "test",
    name: "test",
    onChange: onChangeMock,
};

const renderComponent = ({ ...props }: RadioButtonCardGroupProps) =>
    render(<RadioButtonCardGroup {...props} />);

describe("<RadioButtonCardGroup />", () => {
    describe("Basic smoke tests", () => {
        test("should render without throwing", () => {
            const { container } = renderComponent(PROPS);

            expect(container).not.toBeNull();
        });
    });
});

describe("<RadioButtonCardGroup />", () => {
    describe("On change functionality", () => {
        test("should call onChange when item is selected", async () => {
            const { getAllByRole } = renderComponent(PROPS);

            const firstRadioItem = getAllByRole("radio")[0];

            await userEvent.click(firstRadioItem);

            expect(onChangeMock).toBeCalledWith(RADIO_BUTTON_CARDS_MOCK[0].value);
        });
    });
});
