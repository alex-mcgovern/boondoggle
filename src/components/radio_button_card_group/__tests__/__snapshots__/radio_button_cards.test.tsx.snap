// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButtonCardGroup /> Basic smoke tests should match snapshot 1`] = `
<div>
  <div>
    <label
      class="styles_getGroupLabelStyles__qbf7ok0 get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_color_text_low_contrast__eiij7b1k get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_fontSize_body_sm__eiij7b22 get_sprinkles_fontWeight_normal__eiij7b2h get_sprinkles_gap_space_1_mobile__eiij7bcf get_sprinkles_marginBottom_space_1__eiij7b4n"
      for="test"
      id="label-test"
    >
      test
    </label>
    <fieldset
      aria-labelledby="label-test"
      class="get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_flexWrap_wrap__eiij7b1y get_sprinkles_gap_space_2_mobile__eiij7bci"
      id="test"
      name="test"
    >
      <label
        class="styles_radioButtonCardLabelStyles__1bqsrmi2 a11y_a11yFocus__r8he5t2 styles__1bqsrmi1 get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_aspectRatio_square__eiij7b6 get_sprinkles_border_border_default__eiij7br get_sprinkles_borderRadius_sm__eiij7b10 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_gap_space_4_mobile__eiij7bco get_sprinkles_justifyContent_center__eiij7b3h get_sprinkles_padding_space_4_mobile__eiij7bfc get_sprinkles_width_100%_mobile__eiij7bn3 get_sprinkles_position_relative_mobile__eiij7bml"
        for="test-radio_item_1"
      >
        <input
          class="styles_radioButtonCardInputStyles__1bqsrmi0"
          id="test-radio_item_1"
          name="test"
          type="radio"
          value="radio_item_1"
        />
        <div
          class="get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_flexDirection_column_mobile__eiij7bbr get_sprinkles_justifyContent_center__eiij7b3h get_sprinkles_textAlign_center__eiij7baj"
        >
          <div
            class="get_sprinkles_color_text_low_contrast__eiij7b1k get_sprinkles_marginBottom_space_4__eiij7b4q"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-building-columns fa-xl styles_icon__1ej8lf00 get_sprinkles_color_currentColor__eiij7b1o get_sprinkles_flexShrink_0__eiij7b1w"
              data-icon="building-columns"
              data-prefix="fasr"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M48 192H0V144L256 0 512 144v48H464 48zm366.1-48L256 55.1 97.9 144H228.3c-2.7-4.7-4.3-10.2-4.3-16c0-17.7 14.3-32 32-32s32 14.3 32 32c0 5.8-1.6 11.3-4.3 16H414.1zM64 224h48V384h64V224h48V384h64V224h48V384h64V224h48V384h8 24v48H456 56 32V384H56h8V224zM24 464H488h24v48H488 24 0V464H24z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_gap_space_2_mobile__eiij7bci"
          >
            <div
              class="get_sprinkles_color_text_low_contrast__eiij7b1k"
            >
              Radio item 1
            </div>
          </div>
          <div
            class="get_sprinkles_fontSize_body_xs__eiij7b24 get_sprinkles_lineHeight_body_xs__eiij7b3q get_sprinkles_color_text_low_contrast__eiij7b1k"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          </div>
        </div>
      </label>
      <label
        class="styles_radioButtonCardLabelStyles__1bqsrmi2 a11y_a11yFocus__r8he5t2 styles__1bqsrmi1 get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_aspectRatio_square__eiij7b6 get_sprinkles_border_border_default__eiij7br get_sprinkles_borderRadius_sm__eiij7b10 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_gap_space_4_mobile__eiij7bco get_sprinkles_justifyContent_center__eiij7b3h get_sprinkles_padding_space_4_mobile__eiij7bfc get_sprinkles_width_100%_mobile__eiij7bn3 get_sprinkles_position_relative_mobile__eiij7bml"
        for="test-radio_item_2"
      >
        <input
          class="styles_radioButtonCardInputStyles__1bqsrmi0"
          id="test-radio_item_2"
          name="test"
          type="radio"
          value="radio_item_2"
        />
        <div
          class="get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_flexDirection_column_mobile__eiij7bbr get_sprinkles_justifyContent_center__eiij7b3h get_sprinkles_textAlign_center__eiij7baj"
        >
          <div
            class="get_sprinkles_color_text_low_contrast__eiij7b1k get_sprinkles_marginBottom_space_4__eiij7b4q"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-right-arrow-left fa-xl styles_icon__1ej8lf00 get_sprinkles_color_currentColor__eiij7b1o get_sprinkles_flexShrink_0__eiij7b1w"
              data-icon="arrow-right-arrow-left"
              data-prefix="fasr"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18.2 402.9L111 495.7l17 17 33.9-33.9-17-17L93.1 409.9l330.9 0 24 0 0-48-24 0-330.9 0 55-55 17-17L131.2 256l-17 17-96 96-17 17 17 17zm411.5-256l17-17-17-17-96-96-17-17L282.9 33.9l17 17 55 55L24 105.9l-24 0 0 48 24 0 330.9 0-55 55-17 17 33.9 33.9 17-17 96-96z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_gap_space_2_mobile__eiij7bci"
          >
            <div
              class="get_sprinkles_color_text_low_contrast__eiij7b1k"
            >
              Radio item 2
            </div>
          </div>
          <div
            class="get_sprinkles_fontSize_body_xs__eiij7b24 get_sprinkles_lineHeight_body_xs__eiij7b3q get_sprinkles_color_text_low_contrast__eiij7b1k"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          </div>
        </div>
      </label>
      <label
        class="styles_radioButtonCardLabelStyles__1bqsrmi2 a11y_a11yFocus__r8he5t2 styles__1bqsrmi1 get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_aspectRatio_square__eiij7b6 get_sprinkles_border_border_default__eiij7br get_sprinkles_borderRadius_sm__eiij7b10 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_gap_space_4_mobile__eiij7bco get_sprinkles_justifyContent_center__eiij7b3h get_sprinkles_padding_space_4_mobile__eiij7bfc get_sprinkles_width_100%_mobile__eiij7bn3 get_sprinkles_position_relative_mobile__eiij7bml"
        for="test-radio_item_3"
      >
        <input
          class="styles_radioButtonCardInputStyles__1bqsrmi0"
          id="test-radio_item_3"
          name="test"
          type="radio"
          value="radio_item_3"
        />
        <div
          class="get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_flexDirection_column_mobile__eiij7bbr get_sprinkles_justifyContent_center__eiij7b3h get_sprinkles_textAlign_center__eiij7baj"
        >
          <div
            class="get_sprinkles_color_text_low_contrast__eiij7b1k get_sprinkles_marginBottom_space_4__eiij7b4q"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-shapes fa-xl styles_icon__1ej8lf00 get_sprinkles_color_currentColor__eiij7b1o get_sprinkles_flexShrink_0__eiij7b1w"
              data-icon="shapes"
              data-prefix="fasr"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M160 224h55.3H360.7 416l-27.4-48L315.6 48.4 288 0 260.4 48.4 187.4 176 160 224zM288 96.7L333.3 176H242.7L288 96.7zM128 304a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm0 208a128 128 0 1 0 0-256 128 128 0 1 0 0 256zM464 320V448H336V320H464zM336 272H288v48V448v48h48H464h48V448 320 272H464 336z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_gap_space_2_mobile__eiij7bci"
          >
            <div
              class="get_sprinkles_color_text_low_contrast__eiij7b1k"
            >
              Radio item 3
            </div>
          </div>
          <div
            class="get_sprinkles_fontSize_body_xs__eiij7b24 get_sprinkles_lineHeight_body_xs__eiij7b3q get_sprinkles_color_text_low_contrast__eiij7b1k"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          </div>
        </div>
      </label>
    </fieldset>
  </div>
</div>
`;
